### ISSUE TITLE
Option exclude certain files by pattern when testing recursively

### ISSUE THREAD

**Issue by KylePDavis** [mochajs/mocha#1577]
I would like the ability to provide an exclusion pattern so that I can only test files that match my test file pattern. This would allow data files to coexist with test files so long as they follow a reasonable pattern.

I think this would be simply providing an option that sets `ignore` option in `glob`.

Thoughts? I can make a PR for it real quick if you'd like.

**Comment by alinex** [comment]
You may better make two parallel directories, one for tests and the other
for data files. That's what I did.
Am 04.03.2015 15:52 schrieb "Kyle P Davis" <notifications@github.com>:

**Comment by KylePDavis** [comment]
I could (and have in the past) but I'd rather not have to do that anymore now that I've realized that glob ignore is right there.

It should be pretty easy to add a command line argument to set the glob ignore option. I just wanted to get thoughts and discuss a little bit first.

Any objections to me creating a PR for this?

**Comment by toddbluhm** [comment]
I was just needing this too and was surprised to see mocha did not already have it. I think it would be a great addition. We have a "testApp" inside our test folder that we use specifically for testing. We don't want mocha trying to run tests inside that folder. It would be nice to be able to just add an exclude to our mocha.opts file to exclude that specific path rather than having to explicitly include all the other test sub-folders we have. 

While we could move the testApp out, it really doesn't fit anywhere else in the repo and its got relative pathing to certain files in the test folder.

This I think would be a nice addition. Certainly not critical, but helpful.

**Comment by boneskull** [comment]
@KylePDavis @tawdle 

If I needed this, I think I'd probably just

```sh
$ mocha $(find test/ ! -path '*testApp*')
```

So I'm not really thrilled about the idea.  I'm kind of on the fence about the existence of the `recursive` flag even, since the same thing can be achieved with a shell command, `Makefile`, `Gruntfile.js`, `gulpfile.js`, etc., etc.

**Comment by danielstjules** [comment]
I think @boneskull makes a good point - this is already achievable in a variety of ways. And that's not including the fact that you can structure you directories to avoid this altogether. For example, given this structure:

```
$ tree .
.
└── spec
    ├── fixtures
    ├── integration
    └── unit

4 directories, 0 files
```

You just update your package.json so that you can simply run `npm test`
```
  "scripts": {
    "test": "./node_modules/.bin/mocha spec/unit spec/integration"
  }
```

Or with a structure like the following:
```
$ tree .
.
└── src
    └── models
        ├── user.js
        └── userSpec.js

2 directories, 2 files
```

You could run specs in a similar fashion as @boneskull's example (this will only run files containing `Spec`)
```
  "scripts": {
    "test": "./node_modules/.bin/mocha $(find src -path '*Spec*')"
  }
```

**Comment by boneskull** [comment]
@danielstjules I think that will actually hit any dir with `Spec` in the name. Perhaps you want `-name '*.spec.js'`

**Comment by danielstjules** [comment]
Yup, you're right! Thanks for pointing that out.

Fixed command for the example:
```
  "scripts": {
    "test": "./node_modules/.bin/mocha $(find src -name '*Spec.js')"
  }
```

**Comment by calebthebrewer** [comment]
I think there is still a valid use case for this feature: I tend to group my files by feature, so each test file is next to the files containing the logic they're testing. Naming test files consistently works with the `file` argument or `grep` option but I would like to explicitly ignore things like `node_modules`.

For anyone looking - [gulp-mocha](https://www.npmjs.com/package/gulp-mocha) can achieve this, I just hate to include `gulp` where I don't have to.

**Comment by KrisSiegel** [comment]
I'd like to take a moment and +1 this idea. Unlike typical JavaScript projects I like to follow the way GO does unit testing and I include a spec file next to every single file that gets tested. For example a directory in one of my projects may look like the following:

```
main.js
main.spec.js
utilities.js
utilities.spec.js
```

I find this type of organization makes it far easier to find test versus digging into a single directory of test files; you always expect the test to sit right next to the file that it tests and my build script scripts all .spec.js files out in the deployed version.

Due to my non-standard layout what I'd like to be able to do is run all unit tests with all of my folders excluding the node_modules directory.

I don't have a Windows box near me at the moment but I'm not sure if that find syntax would work in Windows or not; an exclude option would be far more intuitive anyway in my opinion. Plus almost all unit test frameworks include a way of ignoring files anyway so parity would be nice.

**Comment by op1ekun** [comment]
Hi, all my tests support `*.test.js` format. They are all in `tests` directory. Currently I have couple of tests that I want to exclude them and still retain the default path for mocha which is `./test/**/*.js`. How do I do that? 

Creating two separate directories will not work here, and you have to agree that moving  tests between to directories would create a lot of noise in VCS.

@calebthebrewer Thankfully I'm using gulp in the project ;) thanks for the hint!

**Comment by datatypevoid** [comment]
+1 @calebthebrewer

**Comment by datatypevoid** [comment]
Angular 2.0 and Polymer have me in component mode so I agree with @KrisSiegel. Keeping all of your code in bundles keeps a clean, modular, easy-to-maintain file hierarchy.

**Comment by datatypevoid** [comment]
Is there any foreseeable problem to loading the tests using the `q` promise library in a fashion such as this: 

    import q from 'q';

    let q1 = q.defer();

    let q2 = q.defer();

    let q3 = q.defer();

    import authRouterTest from './_authentication.router.e2e';

    import productRouterTest from './_product.router.e2e';
 
    import productModelTest from './product.model.spec';

    productModelTest(q1)

	.then((res, err) => {

		console.log(res);

		authRouterTest(q2);
	});

If not, I think this is the way I will approach it.

**Comment by jeffbski** [comment]
We should still have an ignore for the glob. Using bash commands is not a cross platform solution.

**Comment by godspeedelbow** [comment]
+1 

I really would like to exclude `node_modules`. `Exclusion` seems like a pretty handy companion to `recursive`.

**Comment by isiahmeadows** [comment]
@godspeedelbow Why would that particular directory need explicitly ignored? I don't normally hear about `node_modules` being in a test directory.

FWIW, I could really use this as well. I've got a couple projects where the tests scale far more than the library itself, and this would enable me to better organize them (fixtures belong in a test directory IMHO).

**Comment by KrisSiegel** [comment]
Ignoring ```node_modules``` becomes important when you're either doing something unexpected (like me where I keep spec files next to the file it tests versus keeping them in a single folder, away from the code it tests) or if you want to run unit tests against a main project and some of its included projects (perhaps they're internal libraries, not in the npm repo, that you would like to test) but you don't want to run the unit tests within their own dependencies.

This issue has been around for almost a full year now so I don't have much hope of it being implemented. It's important, in my opinion, to implement because the workarounds either require platform specific terminal commands or you explicitly specify patterns for each directory you want to test (so additional directories need to be added to the testing script). I would fork it and just do it myself but using the hacky method of specifying each directory and a pattern is just easier than finding the time to implement it into mocha.

**Comment by isiahmeadows** [comment]
@KrisSiegel You could always do `src/**/*.js` and `src/**/*.spec.js`. I see that more often in Go and Rust projects than starting in the project root. Matter of fact, the latter is relatively uncommon.

Although I think they're just waiting on someone to just take the time to do it themselves. It doesn't seem like a difficult patch to create (unless that logic is a spaghetti code state machine mess).

I'm not as driven to do this personally because I'm actually working on a test framework of my own, intended to be a complete alternative to Mocha, Jasmine, Tape, etc., although simpler and more powerful at its core. It's still too much a work in progress for even toy projects ATM, though.

**Comment by isiahmeadows** [comment]
I'm using Mocha to effectively bootstrap the framework. I used Chai initially for the assertions until I got those stabilized (they're practically API locked now, since I'm using the framework's assertions to test itself).

**Comment by KrisSiegel** [comment]
Yeah that's a workaround I just don't see many JavaScript projects structured that way though the same could be said about pairing the spec and source file as well. I essentially use that pattern though when I come into projects that are already pretty large with their own custom build systems it's not always easy to change it up.

Exclude is a pretty basic pattern a lot of utilities have. I think mocha needs it :). Though I guess my impression, since this issue was closed based around platform specific workarounds, that it was not desired versus awaiting a patch. If this is something the mocha folks would actually want then if it's not done in a few weeks I could certainly crank out a patch.

**Comment by isiahmeadows** [comment]
@KrisSiegel I'm pretty sure that if someone actually cranked out a patch, it would probably get merged, as long as it doesn't involve some unintuitive, special syntax or extra flag. FWIW, there's yet to be a PR for *this*, and the reason this issue was closed was pretty weak if you consider the  [platform-dependent](https://github.com/mochajs/mocha/issues/1577#issuecomment-166388572). And I don't see very many Windows users that are willing to install GNU `find` just to run a few tests.

It may be closed, but a similar situation is already happening to Node with [Web workers](https://html.spec.whatwg.org/multipage/workers.html), although Node's implementation will slightly deviate ([closed issue](https://github.com/nodejs/node/issues/180), [PR in progress](https://github.com/nodejs/node/pull/2133)).

**Comment by adambuczynski** [comment]
+1 for what it's worth. A project I've built doesn't require an `app` or `src` folder, but instead has a variable amount of named folders with different interface implementations. I would also like to keep my tests bundled for each interface, and not be forced to use a single folder for tests.

Being able to specify files to ignore with a `.mochaignore` or other option would be ideal, as that way I could just run it with a `**/*.spec.js` glob and not worry that it might include tests from the `node_modules` folder.

Almost all other build tools allow this, we have `.npmignore`, `.gitignore`, `.jshintignore`, and jscs provides an option to configure it via `.jscsrc`.

**Comment by godspeedelbow** [comment]
@isiahmeadows like others mentioned, I'm stuck with a folder structure I can't easily change, therefore `services/`, `routes/`, etc. are in the root of the project just like `node_modules` is. I got inspired to put test files next to the files they test. This works really great, but I need a third tool (be it command line, or gulp, or something else) to exclude `node_modules` so that only my own tests get tested.

I'd love to implement it myself in mocha, if I knew where to start :)

**Comment by danielstjules** [comment]
@godspeedelbow @adambuczynski A temporary solution :)
```
  "scripts": {
    "test": "mocha $(find . -name '*.spec.js' ! -ipath '*node_modules*')"
  }
```

**Comment by godspeedelbow** [comment]
Hi @danielstjules thanks! Tried it, but for some reason `find . -name '*.spec.js' ! -ipath '*node_modules*'` only finds one test file.

**Comment by danielstjules** [comment]
Did you update `'*.spec.js'` to fit the pattern your tests follow? E.g. to match _all_ js files, you'd use: 

```
  "scripts": {
    "test": "mocha $(find . -name '*.js' ! -ipath '*node_modules*')"
  }
```

**Comment by adambuczynski** [comment]
@danielstjules cheers, but that wouldn't work on windows I believe, and while for this app it's not a problem, I am hesitant to put it in. 

In the mean time, I've resorted to moving the app code and tests to an `app` subfolder so that I can target that folder for the tests and not worry about `node_modules` or other folders.

**Comment by godspeedelbow** [comment]
@danielstjules yeah, all test files are in format of `*.spec.js`. For some reason it didn't work earlier, but now when it does without problems. Thanks for this *NIX work around.

**Comment by boneskull** [comment]
I want to see this done for an upcoming major; specifically supporting `.mochaignore` seems reasonable to me

**Comment by adambuczynski** [comment]
@boneskull would you also consider implementation of a more "standard" `.mocharc` for passing options, which can be put in the root of the project, rather than having to use a `mocha.opts` in the `test` folder?

That would make configuring Mocha much easier, and in line with the way it's done with other tools out there. Plus, it wouldn't force us to have a `test` folder just for the `mocha.opts`. (We put all our tests alongside the modules they test).

**Comment by boneskull** [comment]
@adambuczynski absolutely.  I am no fan of `mocha.opts`

**Comment by jeffbski** [comment]
@adambuczynski great comment. I agree, I would like to have a standard .mocharc option file too, I no longer use a test directory either, since it is so nice to keep tests next to the code.

**Comment by inf3rno** [comment]
+1, I think exclude is really needed.

**Comment by GRUBES** [comment]
@adambuczynski @boneskull [Documentation](https://mochajs.org/#usage) indicates you can use `--opts` to specify any path to your opts file, though it doesn't give any examples or details.

**Comment by adambuczynski** [comment]
Thanks @GRUBES. In the end, I kept using the `test` folder because I place my test setup helpers in there as well, but it's good to know it's possible.

**Comment by inf3rno** [comment]
Glob has an ignore option, so it should not be a big deal to just add an exlude option and forward it to glob ignore.

> ignore Add a pattern or an array of glob patterns to exclude matches. Note: ignore patterns are always in dot:true mode, regardless of any other settings.

**Comment by isiahmeadows** [comment]
BTW, this would've probably been resolved much earlier had someone actually sat down and wrote the patch.

**Comment by inf3rno** [comment]
@isiahmeadows  Ye, but that won't be me. :D

**Comment by tyrsius** [comment]
So does this still need a PR? I would be happy to add this, since I would prefer to use test file colocation instead of the multiple directories for `test` and `test-integration`

**Comment by ScottFreeCode** [comment]
#2036, which is for basically the same general sort of thing only a file rather than a commandline option, has a PR #2173

**Comment by creeperyang** [comment]
Maybe it's nice to add option like `--exclue`.

Or support something like `mocha test/*.js!test/_*.js`

**Comment by ScottFreeCode** [comment]
Figured Glob out: `mocha "./{,!(node_modules)/**/}*.test.js"` to get all \*.test.js files except in node_modules, and `mocha "./test/**/!(notThisOne).js"` to get everything in the test folder and subfolders except notThisOne.js

See also:
https://github.com/isaacs/node-glob#glob-primer
https://github.com/isaacs/node-glob/issues/62

**Comment by mdumouchel** [comment]
@ScottFreeCode 

When running in terminal I get 

```
mocha "./{,!(node_modules)/**/}*.test.js"
-bash: !: event not found
```

Should this character be escaped?

**Comment by isiahmeadows** [comment]
@mdumouchel Use single quotes.

**Comment by mdumouchel** [comment]
Still getting error
```
node_modules/mocha/lib/utils.js:630
        throw new Error("cannot resolve path (or pattern) '" + path + "'");
              ^
Error: cannot resolve path (or pattern) './{,!(node_modules)/**/}*.test.js'
```

I ended up just using the find command

```
mocha $(find . -type d -name node_modules -prune -o -name '*.test.js')
```

**Comment by isiahmeadows** [comment]
@mdumouchel If you're using version 2.x (i.e. what's on npm), it's not going to work, anyways. Support will start in 3.x, IIRC.

**Comment by ScottFreeCode** [comment]
That's weird, I thought I tried in Bash and was able to use double-quotes. Does anyone know a syntax/escaping that will work both on Windows and in Bash?

**Comment by ScottFreeCode** [comment]
Also, I'm pretty sure the "cannot resolve path" error means that the path doesn't match any files in your project; the path I gave worked in the current version when I had test files with names ending in ".test.js" alongside the corresponding source files, unless I screwed up copying it from my test to the issue comment...

**Comment by boneskull** [comment]
See #2173

**Comment by yvele** [comment]
**Eslint** does a great job ignoring files with `--ignore-path` (and `.eslintignore`).
See: http://eslint.org/docs/user-guide/command-line-interface#ignoring-files

Something similar for mocha would be awesome :two_hearts:

**Comment by boneskull** [comment]
OK, so I relent.  This is just not user-friendly.  

This is what I think we should do:

1.  Support `--exclude <glob-or-path>`
2.  Support *multiple* instances of `--exclude`
3.  Combine all `--exclude` options with all non-option arguments into a single list (basically what [globby](https://npmjs.com/package/globby) does)
4.  Load those files

Currently we support *n* non-option arguments, which can all be globs, but this is only *additive*; this doesn't work like you'd expect:

```shell
$ mocha 'src/**/*.spec.js' '!src/forbidden/**.spec.js'
```

So the above should work, and `--exclude` is really just sugar for `!`.

**Comment by boneskull** [comment]
In addition, `.mochaignore` (#2036) support sounds good, but is a separate issue.  There should be a 3p module we can pull in to have `.gitignore`-like behavior; I'm sure whatever ESLint is doing is good enough for our purposes.

**Comment by boneskull** [comment]
...and if it isn't clear, don't use unquoted globs on the command line; see #2355

**Comment by Tiendq** [comment]
Surprisingly there's still no option like `--ignore-path`, besides putting all tests into a `tests` folder, there's no reason why we can't put tests along with modules.

**Comment by masmau** [comment]
+1 
ignore files or directories by pattern is a very useful feature

**Comment by halt-hammerzeit** [comment]
@isiahmeadows 
> You could always do src/**/*.js and src/**/*.spec.js

No, you couldn't - the `**/*` pattern matching is broken and doesn't work recursively.

**Comment by ScottFreeCode** [comment]
>the `**/*` pattern matching is broken and doesn't work recursively.

Are your paths [quoted](https://github.com/mochajs/mocha/issues/2355#issuecomment-257978896)?

**Comment by nickcodefresh** [comment]
+1

**Comment by matteocontrini** [comment]
![image](https://cloud.githubusercontent.com/assets/2164763/23582220/ebd3361e-0125-11e7-8915-7cc0a927671e.png)

**Comment by izhaknys** [comment]
As describe in https://github.com/zinserjan/mocha-webpack/issues/124:

Inside `src/` there is a file called `server.ts` which fires an express server, and runs at the background. The server is typically up when running coverage, so the port is already in use.

So we don't want this and only this file to be excluded.

**Comment by codebykumbi** [comment]
http://remarkablemark.org/blog/2017/02/07/mocha-glob-pattern/

**Comment by alexanderankin** [comment]
Voicing my opinion that mocha needs an ignore because even after all this time, maintainers still need convincing.

Mocha is canon. why not raise the bar for the cross-platform feature set of the test framework that all other test frameworks aspire to be? We should stop aspiring to leaving everything up to Bash. Its 2017 already.

**Comment by isiahmeadows** [comment]
I'll also point out that Windows people don't have anything equivalent to Bash's `!(glob)` (which is actually a Bash-ism, not even POSIX standard).

**Comment by ScottFreeCode** [comment]
There shouldn't be any Bash dependency in Mocha's globbing as-is, since it is processed through the Glob JS module. (Note: quoting the paths may be required to avoid the shell processing globs differently from how Mocha would, e.g. `**` without the globstar extension or whatever it is that makes that special.)

**Comment by isiahmeadows** [comment]
@ScottFreeCode Do you use `extglob: true` when you use glob? If that's the case, then this can be closed with that syntax as a workaround (since glob/minimatch supports it with that option enabled).

**Comment by ScottFreeCode** [comment]
It looks like we're just [passing the path to glob](https://github.com/mochajs/mocha/blob/73929ad851434a7253252ad054d37c40f9c692b9/lib/utils.js#L667). I am pretty sure that I got negation working through a glob pattern at some point, but that may have been an older version of the module? In any case, we do have [tests for the double-star behavior](https://github.com/mochajs/mocha/blob/075bd51906b828812b320f33cd1c7fa60d1702f1/test/integration/glob.spec.js), so if anyone wants to submit tests to confirm that some negation patterns work too (or if anyone finds any flaws in the globbing tests we already have), that would be great.

However, there's a definite advantage to having an explicit `ignore`/`exclude` option. Multiple advantages, actually:
- less obscure
- easier to avoid clashes with different OSes special characters and quoting rules
- an include list minus an exclude list is simpler in many cases than an include list with negated pieces

I merely wanted to clarify that the status quo isn't meant to be relying on any particular shell. ;^)

**Comment by isiahmeadows** [comment]
> an include list minus an exclude list is simpler in many cases than an include list with negated pieces

True, and I had almost forgotten about that 😄

**Comment by boneskull** [comment]
Anyone coming to this issue:

The workaround is now to use globs, as this should be supported.  If someone wants this *particular* behavior, please send a PR.

**Comment by userpixel** [comment]
I went with `'./{,!(node_modules)}/**/*.test.js'`
instead of , `'./{,!(node_modules)/**/}*.test.js'`

So my package.json looks like this:

```json
"scripts": {
    "lint": "eslint .",
    "utest": "mocha --recursive './{,!(node_modules)}/**/*.utest.js'",
    "intest": "mocha --recursive './{,!(node_modules)}/**/*.intest.js'",
    "test": "npm run lint && npm run utest && npm run intest"
}
```

(`utest` and `intest` is our convention for unit tests and integration tests)

**Comment by sepo-one** [comment]
Does `--exclude` option work ?
Can't find it in [docs](https://mochajs.org/).

**Comment by outsideris** [comment]
@sepo-one I open [PR](https://github.com/mochajs/mocha/pull/3432) to update the document.
You can see all available options vis `mocha -h`.

**Comment by sepo-one** [comment]
Yea, i've been using old version of mocha. Upgraded and `--exclude` option is there.
Docs should be updated anyway i guess.

Thanks @outsideris

**Comment by jtakalai** [comment]
This ought to be available in MochaOptions type as well, then. There is no `exclude` or `ignore` in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mocha/index.d.ts#L2258

Is this a types issue (that "ignore" option exists as it's found in https://github.com/mochajs/mocha/blob/master/example/config/.mocharc.js but not added to type) or does the "non-cli library" not support ignore? I'll submit PR to DefinitelyTyped if you can assure it should work.

**Comment by ahmedelbougha** [comment]
Just for people who may come later to read this, Mocha support `--ignore, --exclude` to ignore files or glob patterns.
Examples:
```
--ignore '**/mytest.js'
--ignore '**/m*.js'
--ignore '**/mytest.*'
```
